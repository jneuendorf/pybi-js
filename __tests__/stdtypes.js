const bytearray = require('../src/bytearray')
const bytes = require('../src/bytes')
const dict = require('../src/dict')
const float = require('../src/float')
const frozenset = require('../src/frozenset')
const int = require('../src/int')
const list = require('../src/list')
const range = require('../src/range')
const set = require('../src/set')
const str = require('../src/str')
const tuple = require('../src/tuple')

const Bytes = require('../src/_bytes')
const {ValueError, NotImplementedError} = require('../src/_errors')

const {createTestCase} = require('./_utils')


describe('bytearray', () => {
    createTestCase('stdtypes', 'bytearray', bytearray, {
        testName: 'without buffer interface objects',
        deserializer: obj => Uint8Array.from(obj),
    })

    test('with buffer', () => {
        expect(
            bytearray(Buffer.from('aÃ¤', 'utf8'))
        ).toEqual(Uint8Array.from([97, 195, 164]))
    })
})


describe('bytes', () => {
    createTestCase('stdtypes', 'bytes', bytes, {
        testName: 'without buffer interface objects',
        deserializer: obj => Bytes.from(obj),
    })

    test('with buffer', () => {
        expect(
            bytes(Buffer.from('aÃ¤', 'utf8'))
        ).toEqual(Bytes.from([97, 195, 164]))
    })

    test('immutability', () => {
        const b = bytes([0, 10])
        expect(b).toBeInstanceOf(Bytes)
        expect(b.length).toBe(2)
        expect(b[0]).toBe(0)
        b[0] = 42
        expect(b[0]).toBe(0)
        b[2] = 1337
        expect(b.length).toBe(2)
        expect(delete b[1]).toBe(false)

        expect(() => b.copyWithin(1, 0)).toThrow(NotImplementedError)
        expect(() => b.fill(3)).toThrow(NotImplementedError)
        expect(() => b.reverse()).toThrow(NotImplementedError)
        expect(() => b.set([3, 4])).toThrow(NotImplementedError)
        expect(() => b.sort()).toThrow(NotImplementedError)
    })
})


describe('dict', () => {
    createTestCase('stdtypes', 'dict', dict, {
        testName: 'basic',
    })

    test('with mapping types', () => {
        const map = new Map([[1, 2], [3, 4]])
        expect(dict(map)).toEqual(map)
        expect(dict(map) === map).toBe(false)
    })

    test('non-string keys', () => {
        // We have include numeric keys (and not have them in the JSON file)
        // because json.dumps({1: 2}) returns '{"1": 2}'.
        expect(dict([[1, 10], [2, 20]])).toEqual(new Map([[1, 10], [2, 20]]))
        const t1 = ['a', 'b']
        const t2 = ['a', 'b']
        expect(dict([[t1, 10], [t2, 20]])).toEqual(new Map([[t1, 10], [t2, 20]]))
    })
})


createTestCase('stdtypes', 'float', float, {
    deserializer(obj) {
        if (typeof(obj) !== 'string') {
            return obj
        }
        const mapping = {
            nan: NaN,
            inf: Infinity,
            '-inf': -Infinity,
        }
        return mapping[obj]
    }
})


describe('frozenset', () => {
    test('construction', () => {
        expect(frozenset()).toBeInstanceOf(Set)
        expect(frozenset().constructor).not.toBe(Set)
        expect(frozenset().size).toBe(0)
        expect(frozenset([1, 'a']).size).toBe(2)
        expect(frozenset([1, 'a']).has('a')).toBe(true)
        expect(frozenset([1, 'a']).has('b')).toBe(false)
        expect([...frozenset([1, 'a'])]).toEqual([1, 'a'])
    })

    test('immutability', () => {
        const fs = frozenset()
        expect(() => fs.add(1)).toThrow(TypeError)
        expect(() => fs.clear()).toThrow(TypeError)
        expect(() => fs.delete(1)).toThrow(TypeError)
    })
})


createTestCase('stdtypes', 'int', int)


createTestCase('stdtypes', 'list', list)


createTestCase('stdtypes', 'range', range)
