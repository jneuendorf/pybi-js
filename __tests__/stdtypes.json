{
    "bytearray": [
        [],
        ["asdf"],
        ["äsdf", "utf8"],
        [10],
        [[1, 5, 10]],
        [["a", "b"]],
        [[100, 500, 1000]]
    ],
    "bytes": [
        [],
        ["asdf"],
        ["äsdf", "utf8"],
        [10],
        [[1, 5, 10]],
        [["a", "b"]],
        [[100, 500, 1000]]
    ],
    "dict": [
        [],
        [[]],
        [["a", 1], ["b", 2]],
        [[["a", 1], ["b", 2]]],
        [[1, 2, 3]],
        [{}],
        [{"a": 1, "b": 2}],
        [""],
        ["asdf"],
        [1],
        [1, 2],
        [null]
    ],
    "float": [
        [],
        [0],
        [10],
        [1.5],
        ["1"],
        [" 1"],
        [" -1"],
        ["1 "],
        ["1.5"],
        [" 1.5"],
        ["1.5 "],
        ["-1.5 "],
        ["+1.5 "],
        ["1.5a"],
        ["asdf"],
        ["nan"],
        ["nAN"],
        ["NaN"],
        ["inf"],
        ["INF "],
        ["InF"],
        [" InFiNity"],
        ["Infinity"],
        ["infinity"],
        ["-inf"],
        ["-INF "],
        ["-InF"],
        ["-InFiNity"],
        [" -Infinity"],
        ["-infinity"],
        [null],
        [[]],
        [{}]
    ],
    "int": [
        [],
        [0],
        [10],
        [1.5],
        [10, 2],
        [1.5, 2],
        [1.5, 1],
        [1.5, -1],
        [1.5, 40],
        ["1"],
        ["1", 0],
        [" 1"],
        [" -1"],
        ["1 "],
        ["1 ", 2],
        ["1a"],
        ["1.5a"],
        ["010"],
        ["-010"],
        ["+010"],
        ["010", 0],
        ["-010", 0],
        ["010", 2],
        ["010", 8],
        ["010", 16],
        ["0x010", 16],
        ["0x010", 0],
        ["0x010"],
        ["-0x010", 16],
        ["-0x010", 8],
        ["-0x010", 2],
        ["-0x010", 0],
        ["-0x010"],
        ["0X010", 16],
        ["0X010", 0],
        ["0X010"],
        ["-0X010", 16],
        ["-0X010", 0],
        ["-0X010"],
        ["0b010", 2],
        ["0b010", 0],
        ["0b010"],
        ["-0b010", 2],
        ["-0b010", 0],
        ["-0b010"],
        ["0B010", 2],
        ["0B010", 0],
        ["0B010"],
        ["-0B010", 2],
        ["-0B010", 0],
        ["-0B010"],
        ["0o010", 8],
        ["0o010", 0],
        ["0o010"],
        ["-0o010", 8],
        ["-0o010", 0],
        ["-0o010"],
        ["0O010", 8],
        ["0O010", 0],
        ["0O010"],
        ["-0O010", 8],
        ["-0O010", 0],
        ["-0O010"],
        ["0b00hh", 32],
        ["asdf"],
        ["nan"],
        ["inf"],
        ["infinity"],
        ["-inf"],
        ["-infinity"],
        [null],
        [[]],
        [{}],
        ["1", 2, 3],
        [{}],
        ["  "]
    ],
    "range": [
        [5],
        [-5],
        [1, 10],
        [1, 10, 2],
        [1, 10, -1],
        [-1, -10],
        [-1, -10, 3],
        [-1, -10, -3],
        [-10, -1],
        [-10, -1, 2],
        [-10, -1, -2],
        [1, 10, 0]
    ]
}
