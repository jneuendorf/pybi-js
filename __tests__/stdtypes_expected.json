{
  "bytearray": [
    [],
    {
      "__error__": {
        "type": "TypeError",
        "message": "string argument without an encoding"
      }
    },
    [
      195,
      164,
      115,
      100,
      102
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0
    ],
    [
      1,
      5,
      10
    ],
    {
      "__error__": {
        "type": "TypeError",
        "message": "'str' object cannot be interpreted as an integer"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "byte must be in range(0, 256)"
      }
    }
  ],
  "bytes": [
    [],
    {
      "__error__": {
        "type": "TypeError",
        "message": "string argument without an encoding"
      }
    },
    [
      195,
      164,
      115,
      100,
      102
    ],
    [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0
    ],
    [
      1,
      5,
      10
    ],
    {
      "__error__": {
        "type": "TypeError",
        "message": "'str' object cannot be interpreted as an integer"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "bytes must be in range(0, 256)"
      }
    }
  ],
  "dict": [
    {},
    {},
    {
      "__error__": {
        "type": "TypeError",
        "message": "dict expected at most 1 argument, got 2"
      }
    },
    {
      "a": 1,
      "b": 2
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "cannot convert dictionary update sequence element #0 to a sequence"
      }
    },
    {},
    {
      "a": 1,
      "b": 2
    },
    {},
    {
      "__error__": {
        "type": "ValueError",
        "message": "dictionary update sequence element #0 has length 1; 2 is required"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "'int' object is not iterable"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "dict expected at most 1 argument, got 2"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "'NoneType' object is not iterable"
      }
    }
  ],
  "float": [
    0.0,
    0.0,
    10.0,
    1.5,
    1.0,
    1.0,
    -1.0,
    1.0,
    1.5,
    1.5,
    1.5,
    -1.5,
    1.5,
    {
      "__error__": {
        "type": "ValueError",
        "message": "could not convert string to float: '1.5a'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "could not convert string to float: 'asdf'"
      }
    },
    "nan",
    "nan",
    "nan",
    "inf",
    "inf",
    "inf",
    "inf",
    "inf",
    "inf",
    "-inf",
    "-inf",
    "-inf",
    "-inf",
    "-inf",
    "-inf",
    {
      "__error__": {
        "type": "TypeError",
        "message": "float() argument must be a string or a number, not 'NoneType'"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "float() argument must be a string or a number, not 'list'"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "float() argument must be a string or a number, not 'dict'"
      }
    }
  ],
  "int": [
    0,
    0,
    10,
    1,
    {
      "__error__": {
        "type": "TypeError",
        "message": "int() can't convert non-string with explicit base"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "int() can't convert non-string with explicit base"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "int() base must be >= 2 and <= 36, or 0"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "int() base must be >= 2 and <= 36, or 0"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "int() base must be >= 2 and <= 36, or 0"
      }
    },
    1,
    1,
    1,
    -1,
    1,
    1,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '1a'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '1.5a'"
      }
    },
    10,
    -10,
    10,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 0: '010'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 0: '-010'"
      }
    },
    2,
    8,
    16,
    16,
    16,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '0x010'"
      }
    },
    -16,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 8: '-0x010'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 2: '-0x010'"
      }
    },
    -16,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '-0x010'"
      }
    },
    16,
    16,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '0X010'"
      }
    },
    -16,
    -16,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '-0X010'"
      }
    },
    2,
    2,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '0b010'"
      }
    },
    -2,
    -2,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '-0b010'"
      }
    },
    2,
    2,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '0B010'"
      }
    },
    -2,
    -2,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '-0B010'"
      }
    },
    8,
    8,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '0o010'"
      }
    },
    -8,
    -8,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '-0o010'"
      }
    },
    8,
    8,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '0O010'"
      }
    },
    -8,
    -8,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '-0O010'"
      }
    },
    11534897,
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: 'asdf'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: 'nan'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: 'inf'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: 'infinity'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '-inf'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '-infinity'"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "int() argument must be a string, a bytes-like object or a number, not 'NoneType'"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "int() argument must be a string, a bytes-like object or a number, not 'list'"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "int() argument must be a string, a bytes-like object or a number, not 'dict'"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "int() takes at most 2 arguments (3 given)"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "int() argument must be a string, a bytes-like object or a number, not 'dict'"
      }
    },
    {
      "__error__": {
        "type": "ValueError",
        "message": "invalid literal for int() with base 10: '  '"
      }
    }
  ],
  "list": [
    [],
    [],
    [
      1,
      2,
      3
    ],
    [
      "a",
      "s",
      "d",
      "f"
    ],
    [
      "a",
      "s",
      "d",
      "f"
    ],
    [
      "a",
      "b"
    ],
    {
      "__error__": {
        "type": "TypeError",
        "message": "list expected at most 1 argument, got 2"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "'NoneType' object is not iterable"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "'int' object is not iterable"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "list expected at most 1 argument, got 2"
      }
    }
  ],
  "range": [
    [
      0,
      1,
      2,
      3,
      4
    ],
    [],
    [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9
    ],
    [
      1,
      3,
      5,
      7,
      9
    ],
    [],
    [],
    [],
    [
      -1,
      -4,
      -7
    ],
    [
      -10,
      -9,
      -8,
      -7,
      -6,
      -5,
      -4,
      -3,
      -2
    ],
    [
      -10,
      -8,
      -6,
      -4,
      -2
    ],
    [],
    {
      "__error__": {
        "type": "ValueError",
        "message": "range() arg 3 must not be zero"
      }
    }
  ],
  "set": [
    [],
    [],
    [
      1,
      2,
      3
    ],
    [
      "s",
      "f",
      "d",
      "a"
    ],
    [
      "s",
      "f",
      "d",
      "a"
    ],
    [
      "s",
      "f",
      "d",
      "a"
    ],
    [
      "b",
      "a"
    ],
    {
      "__error__": {
        "type": "TypeError",
        "message": "set expected at most 1 argument, got 2"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "'NoneType' object is not iterable"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "'int' object is not iterable"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "set expected at most 1 argument, got 2"
      }
    }
  ],
  "str": [
    "",
    "1",
    {
      "__error__": {
        "type": "TypeError",
        "message": "str() argument 2 must be str, not int"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "str() argument 2 must be str, not int"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "str() argument 3 must be str, not int"
      }
    },
    "[['1'], ['2', 3]]",
    {
      "__error__": {
        "type": "TypeError",
        "message": "decoding to str: need a bytes-like object, list found"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "str() takes at most 3 arguments (4 given)"
      }
    },
    "None",
    "True",
    "False",
    "",
    "asdf",
    {
      "__error__": {
        "type": "TypeError",
        "message": "decoding str is not supported"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "decoding to str: need a bytes-like object, int found"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "decoding str is not supported"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "str() argument 2 must be str, not None"
      }
    },
    {
      "__error__": {
        "type": "TypeError",
        "message": "str() argument 2 must be str, not int"
      }
    },
    "[]",
    "['a', 'b']",
    "{}",
    "{'a': 1}"
  ]
}